pkg/provisioner/ironic/ironic.go

```diff
@@ -1289,6 +1289,15 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
 }
 
 func (p *ironicProvisioner) setMaintenanceFlag(ironicNode *nodes.Node, value bool) (result provisioner.Result, err error) {
+	if ironicNode == nil {
+		p.log.Info("no node found, already deleted")
+		return result, nil
+	}
+	if ironicNode.Maintenance == value {
+		p.log.Info("maintenance flag already set", "MaintenanceFlag", value)
+		return result, nil
```

Comment:
It's not clear that an empty result is always the right thing to return in these circumstances. That's one reason this is better checked in the caller.

Reply:
currently in case of error, empty result is also returned, see 
https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1308
(well it could be empty)

Reply:
The result is effectively ignored if the error is non-nil, but here you're returning nil for `err` so it matters.

In the places where this is being called, we do something [completely](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1203-L1206) [different](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1349-L1352) if Maintenance mode is not set.

Reply:
> My thoughts where that a case when `ironicNode.Maintenance == value` is a no-op.

If the call is synchronous (is it?) then you could treat it that way if you remove the `result` return value altogether and just return an `error`.

> so just setting `result.Dirty = true` and skipping `nodes.Update(` could be enough?

That would result in an infinite loop if you removed the checks in the callers.

Reply:
hmm, so to sum up, any changes to current behavior would only complicate things even more :)

Reply:
I dug through a week's worth of Slack threads and rediscovered that @dhellmann told me it _is_ synchronous. So ditching the result altogether and just returning an error is an option.

Reply:
@zaneb Refactored PR taking into account your latest comment